File: src/core/zdex_atomic_buyback_quote_port_v2.py
"""Minimal acyclic quote port for a same-occurrence ZDEX buy-and-burn."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Final

from .global_settlement_types_v1 import (
    MAX_DELTA_ATOMS_V1,
    _require_atoms_u128,
    _require_root,
    hash_global_v1,
)
from .zdex_fee_allocation_types_v1 import FEE_BUYBACK_PRINCIPAL_V1
from .zdex_purchase_burn_route_types_v1 import zdex_pool_reserve_principal_v1

ZDEX_ATOMIC_BUYBACK_QUOTE_PORT_SCHEMA_V2: Final = (
    "zenodex/zdex-atomic-buyback-quote-port/v2"
)


@dataclass(frozen=True, slots=True)
class ZDEXAtomicBuybackQuotePortV2:
    """Proof-independent Tokenomics quote-phase output consumed by Spot.

    Journal and receipt roots are excluded to keep the dependency graph
    acyclic. The final route guest authenticates both child journals and checks
    that they commit this exact port root.
    """

    profile_root: str
    route_release_id: str
    command_occurrence_id: str
    global_pre_state_root: str
    producer_module_release_id: str
    consumer_module_release_id: str
    producer_quote_pre_state_root: str
    producer_quote_post_state_root: str
    producer_quote_effect_plan_root: str
    selected_pool_id: str
    quote_asset_id: str
    amount_atoms: int

    def __post_init__(self) -> None:
        self.validate()

    def validate(self) -> None:
        for field_name in (
            "profile_root",
            "route_release_id",
            "command_occurrence_id",
            "global_pre_state_root",
            "producer_module_release_id",
            "consumer_module_release_id",
            "producer_quote_pre_state_root",
            "producer_quote_post_state_root",
            "producer_quote_effect_plan_root",
            "selected_pool_id",
            "quote_asset_id",
        ):
            value = getattr(self, field_name)
            if type(value) is not str:
                raise TypeError(f"ZDEX quote port {field_name} must be exact str")
            _require_root(value, name=f"ZDEX quote port {field_name}")
        amount = _require_atoms_u128(self.amount_atoms, name="ZDEX quote port amount")
        if amount == 0 or amount > MAX_DELTA_ATOMS_V1:
            raise ValueError("ZDEX quote port amount must fit a positive signed effect")
        if self.producer_module_release_id == self.consumer_module_release_id:
            raise ValueError("ZDEX quote port module releases must differ")
        if self.producer_quote_pre_state_root == self.producer_quote_post_state_root:
            raise ValueError("ZDEX quote phase must change Tokenomics state")

    @property
    def source_principal(self) -> str:
        return FEE_BUYBACK_PRINCIPAL_V1

    @property
    def destination_principal(self) -> str:
        self.validate()
        return zdex_pool_reserve_principal_v1(
            pool_id=self.selected_pool_id,
            asset_id=self.quote_asset_id,
        )

    @property
    def port_root(self) -> str:
        return hash_global_v1("zdex-atomic-buyback-quote-port-v2", self.to_canonical())

    def to_canonical(self) -> dict[str, object]:
        self.validate()
        return {
            "schema": ZDEX_ATOMIC_BUYBACK_QUOTE_PORT_SCHEMA_V2,
            "profile_root": self.profile_root,
            "route_release_id": self.route_release_id,
            "command_occurrence_id": self.command_occurrence_id,
            "global_pre_state_root": self.global_pre_state_root,
            "producer_module_release_id": self.producer_module_release_id,
            "consumer_module_release_id": self.consumer_module_release_id,
            "producer_quote_pre_state_root": self.producer_quote_pre_state_root,
            "producer_quote_post_state_root": self.producer_quote_post_state_root,
            "producer_quote_effect_plan_root": self.producer_quote_effect_plan_root,
            "selected_pool_id": self.selected_pool_id,
            "quote_asset_id": self.quote_asset_id,
            "amount_atoms": self.amount_atoms,
        }


__all__ = [
    "ZDEXAtomicBuybackQuotePortV2",
    "ZDEX_ATOMIC_BUYBACK_QUOTE_PORT_SCHEMA_V2",
]

File: zk/global_settlement_abi_v1/src/zdex_atomic_buyback_quote_port_v2.rs
use serde::{Deserialize, Serialize};

use crate::canonical::{hash_global_v1, AbiErrorV1, AbiResultV1, RootV1};
use crate::zdex_fee_allocation_types::FEE_BUYBACK_PRINCIPAL_V1;
use crate::zdex_purchase_burn_types::zdex_pool_reserve_principal_v1;

pub const ZDEX_ATOMIC_BUYBACK_QUOTE_PORT_SCHEMA_V2: &str =
    "zenodex/zdex-atomic-buyback-quote-port/v2";

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ZDEXAtomicBuybackQuotePortV2 {
    pub schema: String,
    pub profile_root: RootV1,
    pub route_release_id: RootV1,
    pub command_occurrence_id: RootV1,
    pub global_pre_state_root: RootV1,
    pub producer_module_release_id: RootV1,
    pub consumer_module_release_id: RootV1,
    pub producer_quote_pre_state_root: RootV1,
    pub producer_quote_post_state_root: RootV1,
    pub producer_quote_effect_plan_root: RootV1,
    pub selected_pool_id: RootV1,
    pub quote_asset_id: RootV1,
    pub amount_atoms: u128,
}

impl ZDEXAtomicBuybackQuotePortV2 {
    pub fn validate(&self) -> AbiResultV1<()> {
        if self.schema != ZDEX_ATOMIC_BUYBACK_QUOTE_PORT_SCHEMA_V2 {
            return Err(AbiErrorV1::InvalidBinding("ZDEX quote port schema"));
        }
        for root in [
            &self.profile_root,
            &self.route_release_id,
            &self.command_occurrence_id,
            &self.global_pre_state_root,
            &self.producer_module_release_id,
            &self.consumer_module_release_id,
            &self.producer_quote_pre_state_root,
            &self.producer_quote_post_state_root,
            &self.producer_quote_effect_plan_root,
            &self.selected_pool_id,
            &self.quote_asset_id,
        ] {
            root.validate("ZDEX quote port root", false)?;
        }
        if self.amount_atoms == 0 || self.amount_atoms > i128::MAX.unsigned_abs() {
            return Err(AbiErrorV1::InvalidBounds("ZDEX quote port amount"));
        }
        if self.producer_module_release_id == self.consumer_module_release_id {
            return Err(AbiErrorV1::InvalidBinding(
                "ZDEX quote port distinct module releases",
            ));
        }
        if self.producer_quote_pre_state_root == self.producer_quote_post_state_root {
            return Err(AbiErrorV1::InvalidBinding(
                "ZDEX quote phase state transition",
            ));
        }
        Ok(())
    }

    pub fn source_principal(&self) -> &'static str {
        FEE_BUYBACK_PRINCIPAL_V1
    }

    pub fn destination_principal(&self) -> AbiResultV1<String> {
        self.validate()?;
        zdex_pool_reserve_principal_v1(&self.selected_pool_id, &self.quote_asset_id)
    }

    pub fn port_root(&self) -> AbiResultV1<RootV1> {
        self.validate()?;
        hash_global_v1("zdex-atomic-buyback-quote-port-v2", self)
    }
}
